src/graph/graphcontext.c

```diff
@@ -45,40 +37,125 @@ GraphContext *GraphContext_New(RedisModuleCtx *ctx, const char *graphname,
 
 	gc->string_mapping = array_new(char *, 64);
 	gc->attributes = raxNew();
-
+	Graph_SetMatrixPolicy(gc->g, SYNC_AND_MINIMIZE_SPACE);
 	QueryCtx_SetGraphCtx(gc);
 
-	// Set and close GraphContext key in Redis keyspace
-	RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+	return gc;
+}
 
-	GraphContext_RegisterWithModule(gc);
+/* This method tries to get a graph context, and if it does not exists, create a new one.
+ * The try-get-create flow is done when module global lock is acquired, to enforce consistency
+ * while BGSave is called. */
+static GraphContext *_GraphContext_Create(RedisModuleCtx *ctx, const char *graphname,
+										  int readWriteFlag, size_t node_cap, size_t edge_cap) {
+	// Create and initialize a graph context.
+	GraphContext *gc = _GraphContext_New(graphname, node_cap, edge_cap);
+	RedisModuleString *rs_name = RedisModule_CreateString(ctx, graphname, strlen(graphname));
+
+	// Acquire lock.
+	assert(pthread_mutex_lock(&_module_mutex) == 0);
+	// Open key with requested permission.
+	RedisModuleKey *key = RedisModule_OpenKey(ctx, rs_name, readWriteFlag);
+	if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+		// In case of non existing value, open key for write mode.
+		key = RedisModule_OpenKey(ctx, rs_name, REDISMODULE_WRITE);
+		// Set value in key.
+		RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+		// Register graph context for BGSave.
+		GraphContext_RegisterWithModule(gc);
+		// Unlock.
+		assert(pthread_mutex_unlock(&_module_mutex) == 0);
+		goto cleanup;
+	} else {
+		// Unlock - key found successfully.
+		assert(pthread_mutex_unlock(&_module_mutex) == 0);
+		if(RedisModule_ModuleTypeGetType(key) != GraphContextRedisModuleType) {
+			goto cleanup;
+		}
+		// Some other call managed to create and set the context, free the local initialized context.
+		GraphContext_Free(gc);
+		gc = RedisModule_ModuleTypeGetValue(key);
+		// Force GraphBLAS updates and resize matrices to node count by default
+		Graph_SetMatrixPolicy(gc->g, SYNC_AND_MINIMIZE_SPACE);
+
+		QueryCtx_SetGraphCtx(gc);
+	}
 cleanup:
-	RedisModule_CloseKey(key);
 	RedisModule_FreeString(ctx, rs_name);
+	RedisModule_CloseKey(key);
 	return gc;
 }
 
+GraphContext *GraphContext_New(RedisModuleCtx *ctx, const char *graphname,
+							   size_t node_cap, size_t edge_cap) {
+	return _GraphContext_Create(ctx, graphname, REDISMODULE_WRITE, node_cap, edge_cap);
+}
+
 GraphContext *GraphContext_Retrieve(RedisModuleCtx *ctx, const char *graphname, bool readOnly) {
 	GraphContext *gc = NULL;
 	RedisModuleString *rs_name = RedisModule_CreateString(ctx, graphname, strlen(graphname));
 	int readWriteFlag = readOnly ? REDISMODULE_READ : REDISMODULE_WRITE;
```

Comment:
I think we're really overcomplicating the `flag` and general nonexistent-key logic here, since in Redis this flag's only differentiator (outside of WATCH and TTL expiry, which we don't care about) is that `RedisModule_OpenKey` returns NULL in read mode if the key doesn't exist. Better IMO to always open in write mode and then populate if `REDISMODULE_KEYTYPE_EMPTY`.